### DISCUSSION TITLE
multiple store locations for collaboration

### DISCUSSION THREAD

**Discussion by yonicd** [ropensci/targets#938] [Help]
Assuming there are a group of collaborators using the same project area. The project area has multiple pipelines that have core common upstream targets (eg data setup). 

If the collaborators want to leverage the common targets on (ec2's with a common mounted folder) and still build out independent models for themselves without rerunning the common area there is a problem.

Would it be possible to define a vector of store locations that targets could use (named vector for aliases)?

eg common (`tar_config_get(store)`), bob, julie, sally

so when i run a target if i dont specify the store location it would write to use default the common path, but if bob has a periphery model that he is building and doesnt want to conflict with julie or sally he would specify the `tar_target(store=bob)` for his specific targets.

alternatively if bob could reset tar_config_set mid pipeline that would be great too (he would appreciate no needing to write out store on every downstream target (bob is lazy like that))

this isnt exactly the same problem that `{gittargets}` is trying to solve for (I think)

**Comment by wlandau** [comment]
I recommend treating that common store the same way as any ordinary external group of files. Except the function calls to identify the path and read the contents will be different. Sketch:

```r
# _targets.R
library(targets)
list(
  tar_target(bob_data_file, tar_path(bob_data_target_name, store = "bob_data_store_path"),
  tar_target(bob_data, {bob_data_file; tar_read_raw("bob_data_target_name", store = "bob_data_store_path")})
)
```

**Reply by yonicd** [reply]
great! thanks for the tip. i've never tried to use `tar_read_raw` before.

**Reply by yonicd** [reply]
I tried using this new script, but it is for reading, not writing.

For a practical example, how would I store the data I am reading in this pipeline into a location other than what is defined in 

```
tar_config_set(store = '/my_path')
```

i would want to have some targets stored in `'/my_other_path'`

```
files_tbl <- tibble::tibble(
  domain = c('adcm', 'adsl', 'adqs')
) |>
  dplyr::mutate(
    study = dplyr::case_when(
      domain %in% 'adcm' ~ 'iss',
      TRUE ~ 'ise'
    ),
    file = purrr::map2_chr(domain, study, ~make_path(domain = .x, program = 'study_program',study = .y, model = 'adam')),
    label = glue::glue('{study}_{domain}')
  )

list(
  tarchetypes::tar_map(
    files_tbl,
    names = label,
    tar_target(files, file,format = 'file'),
    tar_target(data, haven::read_sas(files), format = 'qs', iteration = 'list')
  )
)
```

the script you have assumes there is something in `'/my_other_path'` unless i am confusing something simple

**Reply by wlandau** [reply]
We're getting into tricky territory outside the scope of `targets`.  I do not have plans to natively support this kind of data store segmentation across multiple collaborators. Of course, you're free to write wherever you want in a target and track the output with `format = "file"`, and [`_targets/user/`](https://books.ropensci.org/targets/data.html#local-data-store) may be a good place for that. But it's tricky when it comes to recording the proper metadata along with that if you're writing into a different `_targets/objects/`, and it would be a Pandora's box to try to tackle natively.

If you use S3 buckets with versioning, the data collaboration/versioning problem is reduced to a much smaller `_targets/meta/meta` file collaboration/versioning problem, and you can put that small flat text file under source control and manage the branching/merging with your colleagues along with the code base. And you can send different target data to different buckets. I talked about this at useR! earlier this year:

* https://www.youtube.com/watch?v=LPt-NPD3rAM (first 13 minutes)
* https://github.com/wlandau/user-conf-2022/
* https://wlandau.github.io/user-conf-2022/

Other than that, I designed the data store so it would be easier for third-party tools and techniques to tackle these data issues where `targets` does not have room to broaden its scope.

**Reply by yonicd** [reply]
Thanks for the clarification. I thought I was getting into uncharted territories with my question :)

**Reply by nicki-dese** [reply]
Would unitar help you here? [https://github.com/databio/unitar](https://github.com/databio/unitar)

**Reply by yonicd** [reply]
Yes! Ty!
